### PR TITLE
#985 pack size controller bug871

### LINKED ISSUE #985: packSize controller: placeholder row part 2
Another bug with the pack size controller.
In an outbound shipment, I have a line like this:

Screen Shot 2022-03-07 at 5 26 45 PM

There is a stock line - but the quantity is only on the placeholder row.
the packSize controller shows a big blank. perhaps #975 could be updated, so the controller shows `1` or `Any` in this case?

### PR DESCRIPTION
**Pull request by josh-griffin**
Fixes #985 

- Fixes the problem in the issue: The pack size selected when there is only a placeholder with allocated packs was wrong.
- Also fixes a problem with expired batches: The pack sizes of expired batches were added to the pack size drop down.
- Also fixes a problem with placeholders showing when the status was allocated+.
- Also fixes a problem with placeholders being allocated quantities when the status was allocated+
- Also fixes a problem that was introduced with the above fixes - when there were no stock lines for an item and the status with allocated+, a blank issue line and table were shown.

Still an issue when delivered+ in #984

### CODE REVIEW

File: packages/common/src/intl/locales/en/distribution.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
   "label.units-in-pack-size-of_one": "unit in pack size of",
   "label.units-in-pack-size-of_other": "units in pack size of",
   "messages.cant-delete-invoices": "Can only delete invoices with a status of 'New' or 'Allocated'",
+  "messages.no-stock-available": "There is no stock available",
```
Comment (joshxg) [inline packages/common/src/intl/locales/en/distribution.json:49]: A bit of a placeholder. This is showing when the status is allocated+ and there is no stock.

It was showing no rows and the auto-allocate input etc

image

Maybe better would be to actually offer a solution.. maybe an an 'order more' button or even a list of links

- Order more
- Create inventory adjustment

Or maybe even an input to easily create an invad through this modal

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 2]
```diff
@@ -110,14 +111,20 @@ export const OutboundLineEdit: React.FC<ItemDetailsModalProps> = ({
           allocatedQuantity={getAllocatedQuantity(draftOutboundLines)}
           availableQuantity={sumAvailableQuantity(draftOutboundLines)}
           onChangeQuantity={onAllocate}
+          canAutoAllocate={!!(currentItem && draftOutboundLines.length)}
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:114]: this is now a big conditional mess.

Wondering what to do about the finalised state of the modal and what to do there. I imagine that will be a significant change also, so thought to leave it messy for now.

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/useDraftOutboundLines.tsx [hunk 3]
```diff
@@ -141,16 +141,17 @@ export const useDraftOutboundLines = (
...
         .sort(sortByExpiry);
 
-      const placeholder = lines?.find(
-        ({ type }) => type === InvoiceLineNodeType.UnallocatedStock
-      );
-
-      if (placeholder) {
-        rows.push(
-          createDraftOutboundLine({ invoiceId, invoiceLine: placeholder })
+      if (status === 'NEW') {
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/useDraftOutboundLines.tsx]: Only add the placeholder if the status is not new

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx [hunk 4]
```diff
@@ -6,30 +6,38 @@ import {
   InvoiceLineNodeType,
   renderHookWithProvider,
 } from '@openmsupply-client/common';
+import { createPlaceholderRow } from './useDraftOutboundLines';
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx:9]: slight tidy up and failing tests (which are now not failing!)

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx [hunk 5]
```diff
@@ -1,102 +1,156 @@
...
-      lines
-        .filter(
-          ({ stockLine }) =>
-            (stockLine?.availableNumberOfPacks ?? 0) > 0 && !stockLine?.onHold
-        )
-        .reduce((sizes, { packSize }) => [...sizes, packSize], [] as number[])
-        .sort((a: number, b: number) => a - b)
-    )
-  );
+type PackSizeOption = {
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx:13]: Not completely happy - the different conditions here is outrageous!

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/utils.ts [hunk 6]
```diff
@@ -88,7 +88,7 @@ export const allocateQuantities =
 
     // If there is only one batch row, then it is the placeholder.
     // Assign all of the new value and short circuit.
-    if (draftOutboundLines.length === 1) {
+    if (draftOutboundLines.length === 1 && status === InvoiceNodeStatus.New) {
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/utils.ts:91]: Only auto-assign to the placeholder when the status is new

File: packages/invoices/src/OutboundShipment/api/operations.generated.ts [hunk 7]
```diff
@@ -432,6 +432,7 @@ export const DeleteOutboundShipmentsDocument = gql`
 export const UpsertOutboundShipmentDocument = gql`
     mutation upsertOutboundShipment($storeId: String!, $input: BatchOutboundShipmentInput!) {
   batchOutboundShipment(storeId: $storeId, input: $input) {
+    __typename
```
Comment (joshxg) [inline packages/invoices/src/OutboundShipment/api/operations.generated.ts:435]: offroad fixing outbound shipment updates causing flickering

File: packages/invoices/src/utils.ts [hunk 8]
```diff
@@ -93,17 +93,17 @@ export const getStatusTranslator =
 
 export const isOutboundDisabled = (outbound: OutboundRowFragment): boolean => {
   return (
-    outbound.status !== InvoiceNodeStatus.Shipped &&
-    outbound.status !== InvoiceNodeStatus.Verified &&
-    outbound.status !== InvoiceNodeStatus.Delivered
+    outbound.status === InvoiceNodeStatus.Shipped ||
```
Comment (joshxg) [inline packages/invoices/src/utils.ts:96]: Wrong outbound disabling

File: packages/common/src/intl/locales/en/distribution.json (hunk 1 above)
Comment (mark-prins) [reply packages/common/src/intl/locales/en/distribution.json:49]: Good idea - have added as a possible requirement for v1+

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx (hunk 2 above)
Comment (mark-prins) [reply packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:114]: a small suggestion is to pull the duplicated `!!(currentItem && draftOutboundLines.length)` out to a variable (`canAutoAllocate`?) just to help readability

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx [hunk 9]
```diff
@@ -293,6 +268,15 @@ describe('usePackSizeController', () => {
     expect(result.current.selected).toEqual({ label: '1', value: 1 });
   });
 
+  it('xxx', () => {
```
Comment (mark-prins) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx:271]:
```suggestion
  it('has an initial value of 1 when only the placeholder row has assigned packs', () => {
```

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx (hunk 4 above)
Comment (mark-prins) [reply packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx:9]: the number of tests shows how complex this thing is 😮  no wonder there were some edge case bugs!

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx [hunk 10]
```diff
@@ -1,102 +1,141 @@
...
-    },
-    [options, setSelected]
+  // Valid pack sizes are the pack size of a line which
+  // - is a placeholder and has allocated stock.
+  // - is a placeholder and is the only line.
+  // - No other placeholders.
+  // - is not on hold.
+  // - is not expired.
+  // - has some available stock.
+  const validPackSizes = uniqBy(
```
Comment (mark-prins) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx:84]: 😁  nice - clear and easy to follow, you've made the best of big stack of conditions

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx [hunk 11]
```diff
@@ -1,102 +1,141 @@
...
+        setPackSize(sameAllocated);
+      }
+    }
+  }, [selected, validPackSizes]);
 
-  return { selected, setPackSize, options, packSizes };
+  return {
+    selected: selected && { label: selected?.label, value: selected?.value },
+    setPackSize,
+    options: validPackSizes.map(({ label, value }) => ({ label, value })),
```
Comment (mark-prins) [inline packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx:139]:
```suggestion
    options: validPackSizes
      .sort(sortPackSizes)
      .map(({ label, value }) => ({ label, value })),
```

and this from GitHub copilot 😉  
```
  // Helper to sort the pack sizes by value.
  const sortPackSizes = (a: PackSizeOption, b: PackSizeOption) => {
    if (a.value === b.value) return 0;
    return a.value < b.value ? -1 : 1;
  };
```
a small suggestion, as I've noticed this:

Screen Shot 2022-03-09 at 9 09 38 AM

**Review by mark-prins — APPROVED** [review]
🎉  Oh my! that's a big effort.. hopefully the final stake in the heart of the recurring zombie issue generator. 🤔  though that's for vampires..? whatever - working well, and super well documented and tested

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx (hunk 9 above)
Comment (joshxg) [reply packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.test.tsx:271]: 😁

File: packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx (hunk 11 above)
Comment (joshxg) [reply packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/usePackSizeController.tsx:139]: Ah yes! Thank you and your copilot!
